queries/tact/textobjects.scm

```diff
@@ -0,0 +1,278 @@
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "function.inner" @_start @_end))
+
+; external
+(external_function
+  body: (_)) @function.outer
+
+(external_function
+  body: (function_body
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "function.inner" @_start @_end))
+
+; contract/trait function
+(function
+  body: (_)) @function.outer
+
+(function
+  body: (function_body
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "function.inner" @_start @_end))
+
+; class.inner & outer
+; -------------------
+(struct
+  body: (struct_body
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "class.inner" @_start @_end)) @class.outer
+
+(message
+  body: (message_body
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "class.inner" @_start @_end)) @class.outer
+
+(contract
+  body: (contract_body
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "class.inner" @_start @_end)) @class.outer
+
+(trait
+  body: (trait_body
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "class.inner" @_start @_end)) @class.outer
+
+; attribute.inner & outer
+; -----------------------
+("@name"
+  "("
+  func_name: (func_identifier) @attribute.inner")") @attribute.outer
+
+(contract_attributes
+  ("@interface"
+    "("
+    (string) @attribute.inner
+    ")") @attribute.outer)
+
+(trait_attributes
+  ("@interface"
+    "("
+    (string) @attribute.inner
+    ")") @attribute.outer)
+
+(trait_attributes
+  ("@interface"
+    "("
+    (string) @attribute.inner
+    ")") @attribute.outer)
+
+; loop.inner & outer
+; ------------------
+(while_statement
+  body: (block_statement
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "loop.inner" @_start @_end)) @loop.outer
+
+(repeat_statement
+  body: (block_statement
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "loop.inner" @_start @_end)) @loop.outer
+
+(do_until_statement
+  body: (block_statement
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "loop.inner" @_start @_end)) @loop.outer
+
+(foreach_statement
+  body: (block_statement
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "loop.inner" @_start @_end)) @loop.outer
+
+; conditional.inner & outer
+; -------------------------
+(if_statement
+  consequence: (block_statement
+    "{"
+    (_) @_start @_end
+    (_)? @_end
+    "}")
+  (#make-range! "conditional.inner" @_start @_end)) @conditional.outer
+
+(if_statement
+  alternative: (else_clause
+    (block_statement
+      "{"
+      (_) @_start @_end
+      (_)? @_end
+      "}"))
+  (#make-range! "conditional.inner" @_start @_end)) @conditional.outer
+
+(if_statement) @conditional.outer
+
+; block.inner & outer
+; -------------------
+(_
+  (block_statement) @block.inner) @block.outer
+
+; call.inner & outer
+; ------------------
+(method_call_expression
```

Comment:
Just a nit but it looks like the parser allows a wide node range for `(method_call_expression)`, might be best to form the `@call.(outer|inner)` queries out of just 
```
name: (identifier)
arguments: (argument_list)
```
 pairs?

Reply:
Hmm, I don't know for sure here. For now went with a more involved approach, but if that becomes a bottleneck somehow, then it could be simplified to your form